libraries/AP_HAL_Linux/Poller.cpp

```diff
@@ -0,0 +1,334 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+ * Copyright (C) 2016  Intel Corporation. All rights reserved.
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <algorithm>
+#include <cstddef>
+#include <errno.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <stdint.h>
+#include <sys/epoll.h>
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#include "Poller.h"
+#include "Scheduler.h"
+
+extern const AP_HAL::HAL& hal;
+
+namespace Linux {
+
+uint32_t Poller::to_epoll_events(Poller::Event ev) {
+    // EPOLLWAKEUP prevents the system from hibernating or suspending when
+    // inside epoll_wait() for this particular event.  It is silently
+    // ignored if the process does not have the CAP_BLOCK_SUSPEND
+    // capability.
+    uint32_t op = EPOLLWAKEUP;
+
+    if (ev & Poller::Event::Read) {
+        op |= EPOLLIN;
+    }
+    if (ev & Poller::Event::Write) {
+        op |= EPOLLOUT;
+    }
+    if (ev & Poller::Event::Error) {
+        op |= EPOLLERR;
+    }
+
+    return op;
+}
+
+bool Poller::register_pollable(Pollable *p, const Poller::Event ev) {
+    if (_epfd < 0) {
+        return false;
+    }
+
+    struct epoll_event epev = {0};
+    epev.events = Poller::to_epoll_events(ev);
+    epev.data.ptr = static_cast<void *>(p);
+
+    return epoll_ctl(_epfd, EPOLL_CTL_ADD, p->get_fd(), &epev) == 0;
+}
+
+void Poller::unregister_pollable(const Pollable *p) {
+    if (_epfd >= 0) {
+        epoll_ctl(_epfd, EPOLL_CTL_DEL, p->get_fd(), NULL);
+    }
+}
+
+int Poller::poll(int timeout_ms) const {
+    const int max_events = 16;
+    epoll_event events[max_events];
+    const auto before_wait_us = AP_HAL::micros64();
+    const auto r = epoll_wait(_epfd, events, max_events, timeout_ms);
+    const auto delta_us = AP_HAL::micros64() - before_wait_us;
+    const auto delta_ms = delta_us / 1000;
+    const auto remaining_time_ms = timeout_ms - delta_ms;
```

Comment:
We should treat the case when `delta_ms > timeout_ms`. According to the conversions rules, the type of `remaining_time_ms` will be unsigned (namely, `uint64_t`). This will become a big value when `delta_ms > timeout_ms`, which will mess up with the calculation of `max_time_ms`.